samples/bluetooth/central_past/src/main.c

```diff
@@ -0,0 +1,326 @@
+
+	if (err) {
+		printk("Failed to connect to %s (%u)\n", addr, err);
+
+		bt_conn_unref(default_conn);
+		default_conn = NULL;
+
+
+		bt_err = bt_le_scan_start(BT_LE_SCAN_ACTIVE, NULL);
+		if (bt_err) {
+			printk("Failed to start scan (err %d)\n", bt_err);
+			return;
+		}
+
+		return;
+	}
+
+	if (conn != default_conn) {
+		return;
+	}
+
+	printk("Connected: %s\n", addr);
+
+	k_sem_give(&sem_conn);
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+	int err;
+
+	if (conn != default_conn) {
+		return;
+	}
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Disconnected: %s (reason 0x%02x)\n", addr, reason);
+
+	bt_conn_unref(default_conn);
+	default_conn = NULL;
+
+	k_sem_give(&sem_conn_lost);
+
+	err = bt_le_scan_start(BT_LE_SCAN_ACTIVE, NULL);
+	if (err) {
+		printk("Failed to start scan (err %d)\n", err);
+		return;
+	}
+}
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+
+static void sync_cb(struct bt_le_per_adv_sync *sync,
+		    struct bt_le_per_adv_sync_synced_info *info)
+{
+	char le_addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(info->addr, le_addr, sizeof(le_addr));
+
+	printk("PER_ADV_SYNC[%u]: [DEVICE]: %s synced, "
+	       "Interval 0x%04x (%u ms), PHY %s\n",
+	       bt_le_per_adv_sync_get_index(sync), le_addr,
+	       info->interval, info->interval * 5 / 4, phy2str(info->phy));
+
+	k_sem_give(&sem_per_sync);
+}
+
+static void term_cb(struct bt_le_per_adv_sync *sync,
+		    const struct bt_le_per_adv_sync_term_info *info)
+{
+	char le_addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(info->addr, le_addr, sizeof(le_addr));
+
+	printk("PER_ADV_SYNC[%u]: [DEVICE]: %s sync terminated\n",
+	       bt_le_per_adv_sync_get_index(sync), le_addr);
+}
+
+static void recv_cb(struct bt_le_per_adv_sync *sync,
+		    const struct bt_le_per_adv_sync_recv_info *info,
+		    struct net_buf_simple *buf)
+{
+	char le_addr[BT_ADDR_LE_STR_LEN];
+	char data_str[129];
+
+	bt_addr_le_to_str(info->addr, le_addr, sizeof(le_addr));
+	bin2hex(buf->data, buf->len, data_str, sizeof(data_str));
+
+	printk("PER_ADV_SYNC[%u]: [DEVICE]: %s, tx_power %i, "
+	       "RSSI %i, CTE %u, data length %u, data: %s\n",
+	       bt_le_per_adv_sync_get_index(sync), le_addr, info->tx_power,
+	       info->rssi, info->cte_type, buf->len, data_str);
+}
+
+static struct bt_le_per_adv_sync_cb sync_callbacks = {
+	.synced = sync_cb,
+	.term = term_cb,
+	.recv = recv_cb
+};
+
+void main(void)
+{
+	struct bt_le_per_adv_sync_param sync_create_param;
+	struct bt_le_per_adv_sync *sync;
+	int err;
+	char le_addr[BT_ADDR_LE_STR_LEN];
+
+	printk("Starting Central Periodic Advertising Synchronization Transfer (PAST) Demo\n");
+
+	/* Initialize the Bluetooth Subsystem */
+	err = bt_enable(NULL);
+
+	printk("Connection callbacks register...");
+	bt_conn_cb_register(&conn_callbacks);
+	printk("Success\n");
```

Comment:
Nitpicking: Can this ever fail?  I assume it is just to be consistent with the pattern below.

Reply:
It cannot. I've remove the "Success" prints